### PR TITLE
Groups

### PR DESCRIPTION
**Pull request by philip-cc**
### Description
This PR implements the group feature as outlined in #227. From a user perspective the following is added:
* One can now create groups through the `CreateGroup` endpoint. 
* One can add/remove users to/from groups using the `AddUserToGroup`/`RemoveUserFromGroup` endpoints.
* One can manage access to objects by providing group IDs to the existing `Permissions` endpoints.

Internally, all users are groups, and access (scopes etc) is checked at the group level. 

### Parent Issue
Issue #227 

### Comments for the Reviewers
Are there more E2E test scenarios we should add?

### Type(s) of Change (Split the PR if you check many)
- [x] Added user feature
- [ ] Added technical feature
- [ ] Added tests
- [ ] Refactor
- [ ] Bugfix
- [ ] Updated documentation
- [ ] Infrastructure changes (Terraform, GCP, K8s, other)
- [ ] CI/CD workflow changes

### Testing Checklist
- [x] I have added/updated unit tests for functional changes.
- [ ] I have added/updated integration tests for changes that affect dependent systems.
- [x] I have added/updated end-to-end tests for feature changes.

### General Checklist
- [x] I have pulled in the latest changes from master.
- [x] I have run `make lint`.
- [x] I have successfully run `make tests`.
- [ ] I have updated relevant workflows and deployment tools.
- [ ] I have updated/added relevant documentation (readme, doc comments, etc).
- [x] I have added the license to new source code files.
- [x] I have spent some time looking over the full diff before creating this PR.

### Technical Debt
Issue #241

### CODE REVIEW

File: encryption-service/impl/authn/authn.go [hunk 1]
```diff
@@ -75,16 +72,40 @@ func (ua *UserAuthenticator) NewUser(ctx context.Context, userscopes users.Scope
...
+		return nil, "", ErrAuthStoreTxCastFailed
+	}
+
+	// Create the user iteself
+	userData, pwd, err := ua.newUserData()
+	if err != nil {
+		return nil, "", err
+	}
+
+	err = authStorageTx.InsertUser(ctx, *userData)
```
Comment (bruno-cc) [inline encryption-service/impl/authn/authn.go:97]: dereferences here and below for groups doesn't look nice for me

File: encryption-service/impl/authstorage/authstorage.go [hunk 2]
```diff
@@ -211,18 +211,52 @@ func (storeTx *AuthStoreTx) RemoveUser(ctx context.Context, userID uuid.UUID) er
...
+func (storeTx *AuthStoreTx) InsertGroup(ctx context.Context, group users.GroupData) error {
+	_, err := storeTx.Tx.Exec(ctx, storeTx.NewQuery("INSERT INTO groups (id, data, key) VALUES ($1, $2, $3)"), group.GroupID, group.ConfidentialGroupData, group.WrappedKey)
+	return err
+}
+
+// Get one or more groups' confidential data
+func (storeTx *AuthStoreTx) GetGroupDataBatch(ctx context.Context, groupIDs []uuid.UUID) ([]users.GroupData, error) {
+	batch := &pgx.Batch{}
+	for _, groupID := range groupIDs {
+		batch.Queue(storeTx.NewQuery("SELECT data, key, id FROM groups WHERE id = $1 AND deleted_at IS NULL"), groupID)
```
Comment (bruno-cc) [inline encryption-service/impl/authstorage/authstorage.go:238]: I am not a fan of this. WHERE IN is always a bit painful, but I think is there right tool here. e.g. https://github.com/jackc/pgx/issues/108 maybe there are better ways now.

File: encryption-service/users/users.go [hunk 3]
```diff
@@ -23,10 +36,31 @@ const (
 	ScopeEnd
 )
 
+const baseAppPath string = "/app.Encryptonize/"
```
Comment (bruno-cc) [inline encryption-service/users/users.go:39]: Are the scopes now not for groups?

File: encryption-service/impl/authn/authn.go [hunk 4]
```diff
@@ -213,3 +250,94 @@ func (ua *UserAuthenticator) NewCLIUser(scopes string, authStore interfaces.Auth
...
+
+	var buffer bytes.Buffer
+	enc := gob.NewEncoder(&buffer)
+	err := enc.Encode(confidential)
+	if err != nil {
+		return nil, err
+	}
+
+	data := buffer.Bytes()
+	wrappedKey, ciphertext, err := ua.UserCryptor.Encrypt(data, groupID.Bytes())
```
Comment (bruno-cc) [inline encryption-service/impl/authn/authn.go:268]: I think we need domain separation here

File: encryption-service/impl/authn/authn.go (hunk 1 above)
Comment (philip-cc) [reply encryption-service/impl/authn/authn.go]: Fixed

File: encryption-service/impl/authstorage/authstorage.go (hunk 2 above)
Comment (philip-cc) [reply encryption-service/impl/authstorage/authstorage.go]: Fixed

File: encryption-service/users/users.go (hunk 3 above)
Comment (philip-cc) [reply encryption-service/users/users.go]: Fixed

File: encryption-service/impl/authn/authn.go (hunk 4 above)
Comment (philip-cc) [reply encryption-service/impl/authn/authn.go]: Done

File: encryption-service/common/scopes.go [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+		case Scope_DELETE:
+			scopes |= ScopeDelete
+		case Scope_INDEX:
+			scopes |= ScopeIndex
+		case Scope_OBJECTPERMISSIONS:
+			scopes |= ScopeObjectPermissions
+		case Scope_USERMANAGEMENT:
+			scopes |= ScopeUserManagement
+		default:
+			return 0, fmt.Errorf("CreateUser: Invalid scope %v", scopes)
```
Comment (bruno-cc) [inline encryption-service/common/scopes.go:93]: this error looks outdated

File: encryption-service/common/scopes.go [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+		case "d":
+			scopes |= ScopeDelete
+		case "i":
+			scopes |= ScopeIndex
+		case "o":
+			scopes |= ScopeObjectPermissions
+		case "m":
+			scopes |= ScopeUserManagement
+		default:
+			return 0, fmt.Errorf("CreateUser: Invalid scope %v", string(scope))
```
Comment (bruno-cc) [inline encryption-service/common/scopes.go:119]: this error looks outdated

File: encryption-service/services/authn/group_handler.go [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+// CreateGroup creates a new group with the requested scopes.
+func (a *Authn) CreateGroup(ctx context.Context, request *CreateGroupRequest) (*CreateGroupResponse, error) {
+	scopes, err := common.MapScopesToScopeType(request.Scopes)
+	if err != nil {
+		log.Error(ctx, errors.New("CreateGroup: Invalid scope"), err.Error())
+		return nil, status.Errorf(codes.InvalidArgument, "invalid scope")
+	}
+
+	userID, err := a.UserAuthenticator.NewGroup(ctx, scopes)
```
Comment (bruno-cc) [inline encryption-service/services/authn/group_handler.go:39]: I find userID confusing here

File: encryption-service/services/authz/authz_middleware.go [hunk 8]
```diff
@@ -80,13 +83,48 @@ func (authz *Authz) AuthorizationUnaryServerInterceptor() grpc.UnaryServerInterc
...
+		}
+
+		reqScope, ok := common.MethodScopeMap[methodName]
+		if !ok {
+			err = status.Errorf(codes.InvalidArgument, "invalid endpoint")
+			log.Error(ctx, err, "AuthzMiddleware: Invalid Endpoint")
+			return nil, err
+		}
+
+		// Find the intersection of the user's groups and the object's groups
```
Comment (bruno-cc) [inline encryption-service/services/authz/authz_middleware.go:99]: Thx go

File: encryption-service/impl/authstorage/authstorage.go [hunk 9]
```diff
@@ -191,11 +191,23 @@ func (storeTx *AuthStoreTx) UserExists(ctx context.Context, userID uuid.UUID) (b
...
+// UpdateUser updates an existing user's data
+func (storeTx *AuthStoreTx) UpdateUser(ctx context.Context, protected *common.ProtectedUserData) error {
+	res, err := storeTx.Tx.Exec(ctx, storeTx.NewQuery("UPDATE users SET data = $1, key = $2 WHERE id = $3"), protected.UserData, protected.WrappedKey, protected.UserID)
+	if err != nil {
+		return err
+	}
+	if res.RowsAffected() < 1 {
+		return interfaces.ErrNotFound
+	}
+	return err
```
Comment (bruno-cc) [inline encryption-service/impl/authstorage/authstorage.go:208]: I think a `nil` is more clear here

File: encryption-service/impl/authstorage/authstorage.go [hunk 10]
```diff
@@ -224,6 +236,72 @@ func (storeTx *AuthStoreTx) GetUserData(ctx context.Context, userID uuid.UUID) (
...
+// InsertGroup inserts a group into the auth store
+func (storeTx *AuthStoreTx) InsertGroup(ctx context.Context, protected *common.ProtectedGroupData) error {
+	_, err := storeTx.Tx.Exec(ctx, storeTx.NewQuery("INSERT INTO groups (id, data, key) VALUES ($1, $2, $3)"), protected.GroupID, protected.GroupData, protected.WrappedKey)
+	return err
+}
+
+// RemoveGroup performs a soft delete by setting a deletion date
+func (storeTx *AuthStoreTx) RemoveGroup(ctx context.Context, groupID uuid.UUID) error {
+	now := time.Now()
+	res, err := storeTx.Tx.Exec(ctx, storeTx.NewQuery("UPDATE groups SET deleted_at = $1 WHERE id = $2"), now, groupID)
```
Comment (bruno-cc) [inline encryption-service/impl/authstorage/authstorage.go:264]: I think there should be an `deleted_at IS NULL`

**Review by bruno-cc — COMMENTED** [review]
Starts to look better and better; I only find the group scopes vs user scopes confusing.

File: encryption-service/common/scopes.go (hunk 5 above)
Comment (philip-cc) [reply encryption-service/common/scopes.go]: Fixed

File: encryption-service/common/scopes.go (hunk 6 above)
Comment (philip-cc) [reply encryption-service/common/scopes.go]: Fixed

File: encryption-service/services/authn/group_handler.go [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+	log "encryption-service/logger"
+)
+
+var ErrAuthStoreTxCastFailed = errors.New("Could not typecast authstorage to authstorage.AuthStoreInterface")
+
+// CreateGroup creates a new group with the requested scopes.
+func (a *Authn) CreateGroup(ctx context.Context, request *CreateGroupRequest) (*CreateGroupResponse, error) {
+	scopes, err := common.MapScopesToScopeType(request.Scopes)
+	if err != nil {
+		log.Error(ctx, errors.New("CreateGroup: Invalid scope"), err.Error())
```
[the remark this answers is not in the document]
Comment (philip-cc) [reply encryption-service/services/authn/group_handler.go]: Fixed

File: encryption-service/services/authn/group_handler.go (hunk 7 above)
Comment (philip-cc) [reply encryption-service/services/authn/group_handler.go]: Fixed

File: encryption-service/impl/authstorage/authstorage.go (hunk 9 above)
Comment (philip-cc) [reply encryption-service/impl/authstorage/authstorage.go]: Fixed

File: encryption-service/impl/authstorage/authstorage.go (hunk 10 above)
Comment (philip-cc) [reply encryption-service/impl/authstorage/authstorage.go]: Fixed

File: encryption-service/impl/authstorage/authstorage.go [hunk 12]
```diff
@@ -191,11 +191,23 @@ func (storeTx *AuthStoreTx) UserExists(ctx context.Context, userID uuid.UUID) (b
...
 // InsertUser inserts a user into the auth store
-func (storeTx *AuthStoreTx) InsertUser(ctx context.Context, protected common.ProtectedUserData) error {
+func (storeTx *AuthStoreTx) InsertUser(ctx context.Context, protected *common.ProtectedUserData) error {
 	_, err := storeTx.Tx.Exec(ctx, storeTx.NewQuery("INSERT INTO users (id, data, key) VALUES ($1, $2, $3)"), protected.UserID, protected.UserData, protected.WrappedKey)
 	return err
 }
 
+// UpdateUser updates an existing user's data
+func (storeTx *AuthStoreTx) UpdateUser(ctx context.Context, protected *common.ProtectedUserData) error {
+	res, err := storeTx.Tx.Exec(ctx, storeTx.NewQuery("UPDATE users SET data = $1, key = $2 WHERE id = $3"), protected.UserData, protected.WrappedKey, protected.UserID)
```
Comment (bruno-cc) [inline encryption-service/impl/authstorage/authstorage.go:201]: We could also think about `deleted_at IS NULL` here 🤔

File: encryption-service/services/authz/authz_handler.go [hunk 13]
```diff
@@ -36,21 +36,21 @@ func (a *Authz) GetPermissions(ctx context.Context, request *GetPermissionsReque
 		return nil, err
 	}
 
-	// Grab user ids
-	uids := accessObject.GetUsers()
-	strUIDs := make([]string, 0, len(uids))
-	for uid := range uids {
-		strUIDs = append(strUIDs, uid.String())
+	// Grab group ids
+	groupIDs := accessObject.GetGroups()
```
Comment (bruno-cc) [inline encryption-service/services/authz/authz_handler.go:40]: How should we handle deleted groups?

File: encryption-service/services/authz/authz_middleware.go [hunk 14]
```diff
@@ -80,13 +83,48 @@ func (authz *Authz) AuthorizationUnaryServerInterceptor() grpc.UnaryServerInterc
...
+		}
+
+		groupIDs := make([]uuid.UUID, 0, len(a))
+		for groupID := range a {
+			if _, ok := b[groupID]; ok {
+				groupIDs = append(groupIDs, groupID)
+			}
+		}
+
+		groupDataBatch, err := authz.UserAuthenticator.GetGroupDataBatch(ctx, groupIDs)
```
Comment (bruno-cc) [inline encryption-service/services/authz/authz_middleware.go:113]: I think this part will blow up with deleted groups (as it errors when less groups are found than requested)

Comment (bruno-cc) [reply encryption-service/services/authz/authz_middleware.go:113]: We should also have a this for this case

File: encryption-service/services/authn/group_handler.go [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+	if err := authStorageTx.Commit(ctx); err != nil {
+		log.Error(ctx, err, "Store: Failed to commit auth storage transaction")
+		return nil, status.Errorf(codes.Internal, "error encountered while storing object")
+	}
+
+	return &AddUserToGroupResponse{}, nil
+}
+
+// RemoveUserFromGroup removes a user from a group.
+func (a *Authn) RemoveUserFromGroup(ctx context.Context, request *RemoveUserFromGroupRequest) (*RemoveUserFromGroupResponse, error) {
```
Comment (bruno-cc) [inline encryption-service/services/authn/group_handler.go:106]: Looks to me that a user can be removed from his user group? See other `GetGroupDataBatch` problems

File: encryption-service/impl/authn/authn.go [hunk 16]
```diff
@@ -108,8 +161,13 @@ func (ua *UserAuthenticator) LoginUser(ctx context.Context, userID uuid.UUID, pr
 		return "", errors.New("Incorrect password")
 	}
 
-	accessToken := NewAccessTokenDuration(userID, userData.Scopes, tokenExpiryTime)
+	// Fetch the user's group and extract scopes
+	groupData, err := ua.GetGroupDataBatch(ctx, []uuid.UUID{userID})
```
Comment (bruno-cc) [inline encryption-service/impl/authn/authn.go:165]: Careful 😄

File: encryption-service/impl/authn/authn.go [hunk 17]
```diff
@@ -108,8 +161,13 @@ func (ua *UserAuthenticator) LoginUser(ctx context.Context, userID uuid.UUID, pr
...
+	if err != nil {
+		return "", err
+	}
 
+	accessToken := NewAccessTokenDuration(userID, groupData[0].Scopes, tokenExpiryTime)
```
Comment (bruno-cc) [inline encryption-service/impl/authn/authn.go:170]: Hm, I am not sure if that whole scopes handling is consistent 🤔 , as it's pretty hard to grasp what the scopes are used for users, users' groups and extra groups.

File: encryption-service/impl/authn/authn.go [hunk 18]
```diff
@@ -108,8 +140,17 @@ func (ua *UserAuthenticator) LoginUser(ctx context.Context, userID uuid.UUID, pr
...
+	groupDataBatch, err := ua.GetGroupDataBatch(ctx, userData.GetGroupIDs())
+	if err != nil {
+		return "", err
+	}
+	combinedScopes := common.ScopeNone
+	for _, groupData := range groupDataBatch {
+		combinedScopes = combinedScopes.Union(groupData.Scopes)
+	}
 
+	accessToken := NewAccessTokenDuration(userID, combinedScopes, tokenExpiryTime)
```
Comment (bruno-cc) [inline encryption-service/impl/authn/authn.go:153]: I think we should note that users "keep" their gained scopes for a while, even after group deletion?

**Review by bruno-cc — APPROVED** [review]

**Pull request merged**